src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -164,44 +140,125 @@ public SubstitutionMatrix(final byte[] matrix) {
+        if (refBase <= 0) {
+            throw new IllegalArgumentException(
+                    String.format("CRAM: Attempt to generate a substitution code for invalid reference base '%c'", (char) refBase));
+        }
+        final byte base = baseByCode[refBase][code];
+        if (base == NO_BASE) {
+            // attempt to retrieve a code for a reference base that isn't in the substitution matrix
+            throw new IllegalArgumentException(String.format("CRAM: Attempt to retrieve a substitution base for invalid base '%c'", (char) refBase));
+        }
+        return base;
+    }
+
     /**
      * Return this substitution matrix as a byte array in a form suitable for serialization.
-     * @return
+     * @return the encoded matrix in serialized form
      */
     public byte[] getEncodedMatrix() {
         return encodedMatrixBytes;
     }
 
+    @Override
+    public String toString() {
+        final int DISPLAY_SIZE = BASES_SIZE * (CODES_PER_BASE + 3) * 2;
+        final StringBuilder stringBuilder = new StringBuilder(DISPLAY_SIZE);
+        for (final SubstitutionBase r : BASES) {
+            stringBuilder.append((char) r.getBase());
+            stringBuilder.append(':');
+            for (int i = 0; i < CODES_PER_BASE; i++) {
+                stringBuilder.append((char) baseByCode[r.getBase()][i]);
+            }
+            stringBuilder.append('\t');
+        }
+        // lower case substitutions
+        for (final SubstitutionBase r : BASES) {
+            char lowerCaseBase = Character.toLowerCase((char)r.getBase());
+            stringBuilder.append(lowerCaseBase);
+            stringBuilder.append(':');
+            for (int i = 0; i < CODES_PER_BASE; i++) {
+                stringBuilder.append((char) baseByCode[lowerCaseBase][i]);
+            }
+            stringBuilder.append('\t');
+        }
+        return stringBuilder.toString();
+    }
+
+    // populate a matrix of substitution frequencies from a list of CramCompressionRecords with Substitution features
+    private static long[][] buildFrequencies(final List<CramCompressionRecord> cramRecords) {
+        final long[][] frequencies = new long[SYMBOL_SPACE_SIZE][SYMBOL_SPACE_SIZE];
+        for (final CramCompressionRecord record : cramRecords) {
+            if (record.readFeatures != null) {
+                for (final ReadFeature readFeature : record.readFeatures) {
+                    if (readFeature.getOperator() == Substitution.operator) {
+                        final Substitution substitution = ((Substitution) readFeature);
+                        final byte refBase = substitution.getReferenceBase();
+                        final byte base = substitution.getBase();
+                        if (refBase <= 0 || base <= 0) {
+                            throw new IllegalArgumentException(
+                                    String.format("CRAM: Attempt to generate a substitution code for invalid reference base '%c'", (char) refBase));
```

Comment:
If the values are negative, printing them as a character doesn't seem like a great way to provide feedback to the user. Unless the user entered this value as a character in the first place.

Reply:
Yeah, true that.

Reply:
If this comes from a careless ```char``` -> ```byte``` cast, a negative here could come from a extended asciii characters like 'ä' (with a code 128 or above). So they are actually printable .... the case that would need to worry are control characters (e.g. tabs new-lines) ... ```java.lang.Character``` query method should give you the ability to tell those apart. Perhaps always print the unsigned code and for those that are printable, you can include the printable version between () beside the code.

Reply:
Changed. If this ever happens, its an internal error, not one the user can recover from, so changing this to display as a signed integer value.